rules/scalac/private.bzl

```diff
@@ -0,0 +1,191 @@
+load("@rules_scala_annex//rules:providers.bzl", "ScalaConfiguration")
+load("//rules/common:private/utils.bzl", "strip_margin")
+load("//rules/common:private/utils.bzl", "write_launcher")
+load("//rules/scala:private/import.bzl", "create_intellij_info")
+
+_common_private_attributes = {
+    "_host_javabase": attr.label(
+        default = Label("@bazel_tools//tools/jdk:current_java_runtime"),
+        cfg = "host",
+    ),
+    "_java": attr.label(
+        default = Label("@bazel_tools//tools/jdk:java"),
+        executable = True,
+        cfg = "host",
+    ),
+    "_jar": attr.label(
+        default = Label("@bazel_tools//tools/jdk:jar"),
+        executable = True,
+        cfg = "host",
+    ),
+    "_jar_creator": attr.label(
+        default = Label("@rules_scala_annex//rules/common/third_party/bazel/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/jarhelper:jarcreator_bin"),
+        executable = True,
+        cfg = "host",
+    ),
+    "_java_toolchain": attr.label(
+        default = Label("@bazel_tools//tools/jdk:current_java_toolchain"),
+    ),
+}
+
+scalac_library_private_attributes = _common_private_attributes
+scalac_binary_private_attributes = _common_private_attributes + {
+    "_java_stub_template": attr.label(
+        default = Label("@anx_java_stub_template//file"),
+    ),
+}
+
+def scalac_binary_implementation(ctx):
+    res = _scalac_common(ctx)
+    launcher = ctx.new_file("%s.sh" % ctx.label.name)
+    files = write_launcher(
+        ctx,
+        launcher,
+        res["java_info"].transitive_runtime_deps,
+        main_class = ctx.attr.main_class,
+        jvm_flags = [],
+    )
+    return _format_providers(ctx, res + {
+        "executable": launcher,
+        "files": files,
+    })
+
+def scalac_library_implementation(ctx):
+    res = _scalac_common(ctx)
+    return _format_providers(ctx, res)
+
+def _scalac_common(ctx):
+    name = ctx.label.name
+    jar = ctx.executable._jar
+    java = ctx.executable._java
+    jar_creator = ctx.executable._jar_creator
+
+    output_jar = ctx.actions.declare_file("%s.jar" % name)
```

Comment:
If this is made an implicit output target (`outputs = {"jar": "%{name}.jar"}` and `ctx.outputs.jar`) in the rule, I think it becomes queryable (as well as documented). Though it does have to be written in more places.

I've done this for the non-basic rules.